tls/bignum.c

```diff
@@ -1025,1146 +881,794 @@ int ttls_mpi_sub_int(ttls_mpi *X, const ttls_mpi *A, ttls_mpi_sint b)
-	ttls_mpi U;
+	unsigned long z = 1;
+	TlsMpi U;
 
-	U.n = U.s = (int) z;
+	U.s = 1;
+	U.limbs = U.used = 1;
 	U.p = &z;
 
-	return(mpi_montmul(A, &U, N, mm, T));
+	return __mpi_montmul(A, &U, N, mm, T);
 }
 
-/*
- * Sliding-window exponentiation: X = A^E mod N  (HAC 14.85)
+/**
+ * Sliding-window exponentiation: X = A^E mod N  (HAC 14.85).
+ *
+ * @X	- destination MPI;
+ * @A	- left-hand MPI
+ * @E	- exponent MPI;
+ * @N	- modular MPI
+ * @RR	- speed-up MPI used for recalculations.
+ *
+ * @RR is used to avoid re-computing R * R mod N across multiple calls,
+ * which speeds up things a bit.
  */
 int
-ttls_mpi_exp_mod(ttls_mpi *X, const ttls_mpi *A, const ttls_mpi *E,
-			const ttls_mpi *N, ttls_mpi *_RR)
+ttls_mpi_exp_mod(TlsMpi *X, const TlsMpi *A, const TlsMpi *E, const TlsMpi *N,
+		 TlsMpi *RR)
 {
-	int ret;
-	size_t wbits, wsize, one = 1;
-	size_t i, j, nblimbs;
-	size_t bufsize, nbits;
-	ttls_mpi_uint ei, mm, state;
-	ttls_mpi RR, T, Apos, *W = *this_cpu_ptr(&g_buf);
-	int neg;
-
-	if (ttls_mpi_cmp_int(N, 0) <= 0 || (N->p[0] & 1) == 0)
-		return(TTLS_ERR_MPI_BAD_INPUT_DATA);
-
+	int r = -ENOMEM, neg;
+	size_t i, j, nblimbs, bufsize = 0, nbits = 0, wbits = 0, wsize, one = 1;
+	unsigned long ei, mm, state = 0;
+	TlsMpi T, Apos, *W = *this_cpu_ptr(&g_buf);
+
+	BUILD_BUG_ON(MPI_W_SZ < 6);
+	if (ttls_mpi_cmp_int(N, 0) <= 0 || !(N->p[0] & 1))
+		return -EINVAL;
 	if (ttls_mpi_cmp_int(E, 0) < 0)
-		return(TTLS_ERR_MPI_BAD_INPUT_DATA);
+		return -EINVAL;
 
-	/*
-	 * Init temps and window size
-	 */
-	mpi_montg_init(&mm, N);
-	ttls_mpi_init(&RR); ttls_mpi_init(&T);
+	/* Init temps and window size. */
+	__mpi_montg_init(&mm, N);
+	ttls_mpi_init(&T);
 	ttls_mpi_init(&Apos);
-	memset(W, 0, sizeof(ttls_mpi) * MPI_W_SZ);
+	memset(W, 0, sizeof(TlsMpi) << MPI_W_SZ);
 
 	i = ttls_mpi_bitlen(E);
+	wsize = (i > 671) ? 6
+		: (i > 239) ? 5
+		  : (i >  79) ? 4
+		    : (i >  23) ? 3
+		      : 1;
+
+	j = N->used + 1;
+	if (ttls_mpi_grow(X, j)
+	    || __mpi_realloc(&W[1], j, 0)
+	    || __mpi_realloc(&T, j * 2, 0))
+		goto cleanup;
 
-	wsize = (i > 671) ? 6 : (i > 239) ? 5 :
-			(i >  79) ? 4 : (i >  23) ? 3 : 1;
-
-	if (wsize > TTLS_MPI_WINDOW_SIZE)
-		wsize = TTLS_MPI_WINDOW_SIZE;
-
-	j = N->n + 1;
-	TTLS_MPI_CHK(ttls_mpi_grow(X, j));
-	TTLS_MPI_CHK(ttls_mpi_grow(&W[1],  j));
-	TTLS_MPI_CHK(ttls_mpi_grow(&T, j * 2));
-
-	/*
-	 * Compensate for negative A (and correct at the end)
-	 */
+	/* Compensate for negative A (and correct at the end). */
 	neg = (A->s == -1);
-	if (neg)
-	{
-		TTLS_MPI_CHK(ttls_mpi_copy(&Apos, A));
+	if (neg) {
+		if ((r = ttls_mpi_copy(&Apos, A)))
+			goto cleanup;
 		Apos.s = 1;
 		A = &Apos;
 	}
 
 	/*
 	 * If 1st call, pre-compute R^2 mod N
 	 */
-	if (_RR == NULL || _RR->p == NULL)
-	{
-		TTLS_MPI_CHK(ttls_mpi_lset(&RR, 1));
-		TTLS_MPI_CHK(ttls_mpi_shift_l(&RR, N->n * 2 * biL));
-		TTLS_MPI_CHK(ttls_mpi_mod_mpi(&RR, &RR, N));
-
-		if (_RR != NULL)
-			memcpy(_RR, &RR, sizeof(ttls_mpi));
+	if (!RR || !RR->p) {
```

Comment:
This part makes no sense anymore. Previously, one could pass `NULL` as `_RR`, and in that case this function did the needed calculation. If `_RR` wasn't `NULL`, it was used as `RR`. Now the function requires a pre-computed `RR` and won't work correctly if there is `NULL`. Moreover, in that case `ttls_mpi_lset(RR, 1)` will be called, which will dereference `NULL`.